### PR TITLE
Add Packing Support for Context Parallelism (Ring Attention)

### PR DESCRIPTION
**Pull request by kocchop**
# Description

Enables sequence packing for context parallelism with `ring` strategy using TransformerEngine's DotProductAttention. Includes comprehensive GPU tests for ring attention with packing for sm90+.

- Currently supports packing only for ring attention
- Replaced local sequence reordering with TE `reorder_causal_load_balancing` api
- Currently the load balancing strategy is automatically picked based on the packing config

Current support matrix for context parallelism on GPU:

CP Strategy | Packing | Load Balance Strategy | SWA
-- | -- | -- | --
`all_gather` | ❌ | `dual_chunk_swap` | ❌
`ring` | ✅ (with real data) | `dual_chunk_swap` (non-packed) `striped` (packed) | ✅ (with packing)

# Tests

Added a GPU integration test that works for sm90+.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/MaxText/layers/attention_op.py [hunk 1]
```diff
@@ -1385,18 +1394,26 @@ def cudnn_flash_attention(
 
     # Handle packing configurations
     if self.config.packing and self.config.dataset_type != "synthetic":
+      if using_context_parallelism and not using_load_balanced_ring_cp:
+        raise AssertionError("Packing is only supported for load balanced ring attention with context parallelism.")
```
Comment (richjames0) [inline src/MaxText/layers/attention_op.py]: Nit: AssertionError feels weird here to me. Maybe an argumenterror?

File: src/MaxText/maxtext_utils.py [hunk 2]
```diff
@@ -107,19 +107,24 @@ def get_functional_eval_with_signature(eval_step, data_sharding, state_mesh_shar
...
+  reordered = max_utils.reorder_causal_load_balanced(batch, cp_size, reorder_strategy)
   for _, v in batch.items():
     if isinstance(v, jax.Array):
       reordered = sharding.maybe_shard_with_name(reordered, v.sharding, shard_mode)
       break
   return reordered
 
 
-def get_reorder_callable(cp_size, shard_mode):
+def get_reorder_callable(cp_size, shard_mode, reorder_strategy=0):  # 0=DualChunkSwap, 1=Striped
```
Comment (richjames0) [inline src/MaxText/maxtext_utils.py]: As I read this late at night I imagine you're using an integer here so it's comprehensible by JAX but could this be made into an enum without breaking things (at worse using `.value`?

**Review by richjames0 — APPROVED** [review]
a couple of nits but lgtm

File: src/MaxText/maxtext_utils.py [hunk 3]
```diff
@@ -107,19 +107,24 @@ def get_functional_eval_with_signature(eval_step, data_sharding, state_mesh_shar
   return functional_eval, in_shardings, out_shardings, static_argnums, donate_argnums
 
 
-def shard_reorder_causal_load_balanced(batch, cp_size, shard_mode):
+def shard_reorder_causal_load_balanced(batch, cp_size, shard_mode, reorder_strategy=0):
```
Comment (gobbleturk) [inline src/MaxText/maxtext_utils.py]: Can we make reorder_strategy configurable via base.yml?

Comment (kocchop) [reply src/MaxText/maxtext_utils.py]: done, ptal @gobbleturk

File: src/MaxText/maxtext_utils.py (hunk 2 above)
Comment (kocchop) [reply src/MaxText/maxtext_utils.py]: changed to enum @richjames0

File: src/MaxText/layers/attention_op.py (hunk 1 above)
Comment (kocchop) [reply src/MaxText/layers/attention_op.py]: converted to `ValueError`

File: src/maxtext/utils/max_utils.py [hunk 4]
```diff
@@ -791,13 +791,50 @@ def reorder_sequence(tensor, cp_size: int, seq_dim: int = 1, to_contiguous: bool
   return reordered.reshape(ori_tensor_shape)
 
 
-@partial(jax.jit, static_argnums=1)
-def reorder_causal_load_balanced(batch, cp_size):
-  """Reorders the example batch sequences"""
+@partial(jax.jit, static_argnums=(1, 2))
+def reorder_causal_load_balanced(batch, cp_size, reorder_strategy):
+  """Reorders the example batch sequences
```
Comment (gobbleturk) [inline src/maxtext/utils/max_utils.py]: Great comment explaining the two of them with examples, thank you!

File: src/MaxText/maxtext_utils.py (hunk 3 above)
Comment (gobbleturk) [reply src/MaxText/maxtext_utils.py]: Comment - Terminology/naming is hard, folks have been using the name "striped" to refer to DUAL_CHUNK_SWAP. I guess the string "striped" has to be passed to transformer engine for the other strategy (I would prefer the name "interleaved")...

I highly appreciate your comment with examples of the two strategies to clearly show what they mean in our codebase anyway!

File: src/maxtext/configs/base.yml [hunk 5]
```diff
@@ -881,6 +881,7 @@ dq_reduction_steps: 0 #the number of reduction steps. For now, only 3 or all the
 ### Determine if we want to use load balance for context parallelism
 context_parallel_load_balance: True
 context_parallel_strategy: "all_gather" # "all_gather" or "ring"
+context_parallel_reorder_strategy: "auto" # "auto", "dual_chunk_swap", or "striped"
```
Comment (RissyRan) [inline src/maxtext/configs/base.yml:1041]: Thanks for adding this strategy! Could you help add some explanation here for each from `reorder_causal_load_balanced`?

File: src/maxtext/utils/train_utils.py [hunk 6]
```diff
@@ -186,26 +187,43 @@ def setup_train_loop(config, recorder, devices=None):
...
+            "Please use 'ring' strategy instead."
+        )
 
     # Apply reordering wrapper to data iterators if context parallelism is enabled
     with jax.set_mesh(mesh):
       if context_parallel_size > 1 and config.context_parallel_load_balance:
-        data_iterator = map(maxtext_utils.get_reorder_callable(context_parallel_size, config.shard_mode), data_iterator)
+
+        # Determine load balancing reorder strategy based on whether packing is enabled
+        if config.context_parallel_reorder_strategy == ReorderStrategy.AUTO:
```
Comment (RissyRan) [inline src/maxtext/utils/train_utils.py:251]: Wondering if AUTO usually gives the best result? Or if there will be an compatible issue if user select ReorderStrategy.STRIPED, but without packing?

Trying to understand if we just provides 2 strategies are good enough.

File: src/maxtext/configs/base.yml (hunk 5 above)
Comment (kocchop) [reply src/maxtext/configs/base.yml:1041]: Hi @RissyRan, I was ooo, could not follow up. The explanation for each of the reorder strategies are there with some example in `max_utils.py`. Are you asking to include it in the `base.yml`?

File: src/maxtext/utils/train_utils.py (hunk 6 above)
Comment (kocchop) [reply src/maxtext/utils/train_utils.py:251]: Hi @RissyRan, you can think of `AUTO` as a heuristics for users who may not want to deal with different reordering + cp strategy intricacies to begin with. We plan to add more support for reordering, packing with other CP strategies in future. That's why proposing this. If user has high level confidence and understanding, then they can just choose themselves, otherwise `AUTO` will try to pick based on the CP type, packing etc. configs. Let me know if it makes sense.

File: src/MaxText/maxtext_utils.py (hunk 3 above)
Comment (kocchop) [reply src/MaxText/maxtext_utils.py]: Hi @gobbleturk , so do you want to change the MaxText strategy name to `interleaved`?

File: src/maxtext/configs/base.yml (hunk 5 above)
Comment (RissyRan) [reply src/maxtext/configs/base.yml:1041]: Yes, could you add some short explanation here. Users usually refer to base config instead of checking source codes. Some concise explanation will be great!
